README.md
# Feather

This tool allows you to easily build **interactive seat selection** into your website and is powered by the Ingresso API.

## Demo

[Click to view the demo](https://ingresso-group.github.io/feather-demo/)

### Deploying the Demo

-   To deploy the demo, run the `npm run deploy` command

#### To keep a nice clean release history you can additionally

-   Click on [Releases](https://github.com/ingresso-group/feather-demo/releases)
-   Click on _Draft a new release_
-   Fill in the tag number with the current version of _Feather Demo_ (not the version of _Feather_) as per the Changelog
-   Fill in the _Release Title_
-   Click on _Publish Release_

## Getting Started

1. Create an empty html file

2. Add a container element, where the widget will appear

3. Import our library at the bottom of your page:
   [Get it here](https://storage.googleapis.com/ticketswitch/feather/latest/feather.min.js)

4. Configure the widget and initialise the chart

5. Open in the browser

### A Simple Example

```html
<!DOCTYPE html>

<html>
    <head>
        <style>
            #ingresso-widget {
                width: 500px;
                height: 500px;
            }
        </style>
    </head>

    <body>
        <div id="ingresso-widget"></div>

        <script src="https://storage.googleapis.com/ticketswitch/feather/latest/feather.min.js"></script>
        <script>
            var chartConfig = {
                eventID: "7AB", // demo event
                perfID: "7AB-9",
                selector: "#ingresso-widget",
                token: "<The-X-B2B-Token>", // see the Authentication section below
            };
            var chart = new IngressoSeatingPlan();
            chart.init(chartConfig);
        </script>
    </body>
</html>
```

![Alt text](example.png?raw=true)

> For some more examples of using feather in other JS frameworks see ./examples

## Authentication

### Getting Authenticated

We have a demo user set up, that allows you to create an auth token for our demo so that you can experiment with the seat chart.

#### 1. Get an auth token

```bash
curl https://b2b.ingresso.co.uk/api/b2b/ -u "demo:demopass" -ik
```

This will have a `x-b2b-token` in the response header

#### 2. Add your token to your chart config

```js
var chartConfig = {
    eventID: "7AB", // demo event
    perfID: "7AB-9",
    selector: "#ingresso-widget",
    token: "<The-X-B2B-Token>", // 👈👈👈
};
var chart = new IngressoSeatingPlan();
chart.init(chartConfig);
```

### Authentication Details

In order to reserve tickets, you need to obtain a username and password. After receiving those, you should make a request to <https://b2b.ingresso.co.uk/api/b2b> with your `username` and `password` in the `Authorization` header, as basic auth. The response will have a custom HTTP Header, **x-b2b-token**. You can then pass this token in the configuration object of the widget, and will allow you to purchase orders based on the **transaction_uuid** value.

Each authentication token has a **maximum** lifetime of 4 hours and is valid for all your customers, so it is advisable that you make this call ahead of time and cache the result, in order to improve the UX of your page (as opposed to making it on each page load). However, this token can get invalidated at any time, which means that you should generate a new one very often (once every few seconds is fine).

**Note:** We are most likely going to change this and move to a model of manually-generated tokens that will have an unlimited lifetime and this would remove the complexity of generating and managing them on your side, but in the meantime, you should use the approach presented above.

## Widget Configuration

In order to specify the event/performance to load availability for, as well as to specify where to load the widget from, we use a configuration object.

#### Required configuration parameters:

-   eventID (String)
-   perfID (String)
-   selector (String) (the identifier of the container that will hold the widget)

#### Optional configuration parameters:

-   token (String) (Authentication token, not required to view the demo, but required to purchase; read more about how to obtain this in [Authentication](#authentication)
-   domain (String, default is "https://b2b.ingresso.co.uk" )
-   silenceWarnings (Boolean, default is **false**) (If set to **false**, the widget will provide warnings regarding missing configuration and suggestions on fixing issues)
-   preloaderColor (String) (Hex color e.g **#EC008C**)
-   hasCustomLegend (Boolean, default is **false**) (If you plan on changing the prices/messages displayed, you have to set this to **true** - see the **setLegend** function below)
-   allowControlsOnSmallScreens (Boolean, default is **false**) (Normally, **plus/minus/reset** controls get hidden on small/mobile devices, because in most cases, there isn't enough room or they simply aren't needed. This allows you to override that behaviour)

## Getting information out of the widget

In order to allow the host page to respond to user actions on the seating plan, there are a suite of callbacks available:

-   onChartInitialised
-   onAddSeat
-   onRemoveSeat
-   onEmptyBasket
-   onGoToCheckout
-   onNewAvailabilityData
-   onNewLegendColors
-   onReserveStopped

All of the callbacks will also receive an object containing all the data required to display the right information to the user.

**NOTE:** if you set a custom legend with different prices, **DO NOT** use the basket total that gets returned by these callbacks, as it will not reflect the prices that you've set in the legend. You have to compute your own total, based on the currently selected seats.

## Controlling the widget

There are also some imperative methods used to control the widget, that we can call either before or after initialising it:

-   selectSeatsByIdAndBlock ({ seat_id: string, seat_block: string })
    > You can also pass an array of such seat objects to select multiple seats
-   zoomIn
-   zoomOut
-   resetChart
-   hideControls
-   showControls
-   hideLegend
-   showLegend
-   reserve
-   release
-   disableScrollToZoom
-   disableInitialToZoom
-   changeColorScheme (Array of strings) (Hex colors, e.g ['#462446', '#B05F6D', '#EB6B56', '#FFC153', '#47B39D', '#CDCBA6', '#008891', '#00587A', '#ff0000', '#0F3057'])
-   selectPerformance (String) (ID of desired performance, you get this from the Ingresso API; when this function is called, the widget will automatically make itself visible)
-   setPreloader ({css, html}) - if you set this, it will override the default preloader
-   setLegend (Object) - when new availability data is received, you can take the legend object, change any values within it (prices, messages) and then use this function to send it back to the app. It will immediately update the displayed prices.

Availability data and colours are received via their own callbacks (**onNewAvailabilityData** and **onNewLegendColors**) and are needed in order to display pricing information and colors for the selected seats. (Each seat's data contains a reference to a specific legend item, for price/colour)

## Reserving Seats

It's enough to call the **reserve** method on the library, which will start the reservation process. It will handle edge cases as well (if the selected seats have become unavailable in the meantime, or the back-end has responded with any other error) and will fire the **goToCheckout** callback when the reservation process is complete or **onReserveStopped** if an error has appeared.

When called, **goToCheckout** will contain a **transaction_uuid** property, that represents a unique identifier assigned to that specific order. You can then go on to reserve that order, based on the token, via the Ingresso API.

    Note: the seats requested by the user might become unavailable in the meantime

If the requested seats happen to have become unavailable before the user can reserve them, the widget will display a modal window giving the user a choice between going back and selecting other seats manually or going through with a pre-selected set of seats recommended by our system.

-   If they choose to select different seats manually, the modal closes and the initially-selected seats drop off the seating map.
-   If they choose to proceed with the recommended seats, the modal will also close and the app will make a reservation call for the new seats. When this process finishes, the newly-created reservation will be for the recommended seats, which you can then display on your checkout page (or anywhere you choose).

## Customisation

The widget supports easy customisation of the following attributes:

-   seating plan colors
-   fonts
-   texts
-   look & feel (as long as you provide us with the required CSS, we'll apply it for your integration only)
-   prices (legend and hover tooltip)

However, all of these are changed via our internal control panel, so you have to ask us to do them for you.

The only things that can be changed by your are the color scheme on the seating plan and the preloader color, which you should provide in the configuration object of the widget (see [Configuration](#widget-configuration))

## Optimisations

Even though the widget allows you to initialise it with a performance ID on page load, in most cases you should initialise it just with the event ID and no performance ID, then have it load in the background (If it doesn't have a performance ID when it is initialised, the widget will hide itself).

In most cases, it is best to have a calendar that allows users to choose a performance, which you then pass to the widget.When this happens, the widget will already be ready and will become visible and there should be practically no delay before they see the interactive seating plan with availability loaded (the initial availability that the user sees is cached for a short time and then it updates, which makes for very short loading times).

## Widget preloader

The library creates, shows and hides the preloader for you, transparently.


CHANGELOG.md
## Change log

## [0.2.8] - 2022-11-08

### Updated

-   Update to feather v0.4.8 and pointer to latest gcloud folder
-   Added toggle for hiding currency symbol

## [0.2.7] - 2021-05-06

### Updated

-   Update to feather v0.4.7

## [0.2.6] - 2021-04-14

### Updated

-   Update to feather v0.4.6

## [0.2.5] - 2021-03-12

### Updated

-   Update to feather v0.4.5

## [0.2.4] - 2021-02-18

### Updated

-   Update to feather v0.4.4

## [0.2.3] - 2021-02-12

### Updated

-   Update default perf to be from 2022 instead of 2020

## [0.2.2] - 2019-06-03

### Security

-   Patch security vulnerability in axios

## [0.2.1](https://storage.googleapis.com/ticketswitch/feather/0.2.1/feather.min.js) - 2018-06-29

### Added

-   added a handler for 404 errors: this can be handled in any way is deemed reasonable, but at least they won't show an preloader with an unrelated page underneath anymore

## [0.2.0](https://storage.googleapis.com/ticketswitch/feather/0.2.0/feather.min.js) - 2017-11-30

### Added

-   added possibility to specify a custom legend
-   changed the way the initial performance is selected (removed it from the url and instead passed it as a regular **selectPerformance** call

## [0.1.2](https://storage.googleapis.com/ticketswitch/feather/0.1.2/feather.min.js) - 2017-11-28

### Added

-   added possibility to set a custom preloader

## [0.1.1](https://storage.googleapis.com/ticketswitch/feather/0.1.1/feather.min.js) - 2017-11-23

### Added

-   added setPreloader method
-   added a flag for forcing controls to always show, even on small screens

## [0.1.0](https://storage.googleapis.com/ticketswitch/feather/0.1.0/feather.min.js) - 2017-11-13

### Added

-   force 'choose num tickets' ab test to false

## [0.0.17](https://storage.googleapis.com/ticketswitch/feather/0.0.17/feather.min.js) - 2017-11-06

### Added

-   added functionality to programatically select seats

## [0.0.16](https://storage.googleapis.com/ticketswitch/feather/0.0.16/feather.min.js) - 2017-10-10

### Fixed

-   silenced the 'missing token' warning when silenceWarnings is set to true

## [0.0.15](https://storage.googleapis.com/ticketswitch/feather/0.0.15/feather.min.js) - 2017-10-10

### Added

-   added availabilityURL to config parameters, allowing people to specify a custom endpoing for getting availability data

## [0.0.14](https://storage.googleapis.com/ticketswitch/feather/0.0.14/feather.min.js) - 2017-09-28

### Fixed

-   fixed CSS margin of the iframe

## [0.0.13](https://storage.googleapis.com/ticketswitch/feather/0.0.13/feather.min.js) - 2017-09-27

### Fixed

-   fixed CSS margin of the iframe

## [0.0.12](https://storage.googleapis.com/ticketswitch/feather/0.0.12/feather.min.js) - 2017-09-22

### Fixed

-   fixed infinite refresh loop

## [0.0.11](https://storage.googleapis.com/ticketswitch/feather/0.0.11/feather.min.js) - 2017-09-22

### Fixed

-   fixed issue with Feather not loading when coming back from checkout page

## [0.0.9](https://storage.googleapis.com/ticketswitch/feather/0.0.9/feather.min.js) - 2017-09-14

### Changed

-   Updated the documentation URL that is displayed in the console when Feather is not initialised/used correctly

### Added

-   Added unit tests

## [0.0.9](https://storage.googleapis.com/ticketswitch/feather/0.0.9/feather.min.js) - 2017-09-14

### Changed

-   Updated the documentation URL that is displayed in the console when Feather is not initialised/used correctly

### Added

-   Added unit tests

## [0.0.8](https://storage.googleapis.com/ticketswitch/feather/0.0.8/feather.min.js) - 2017-09-14

### Added

-   added functions: disableScrollToZoom and disableInitialZoom
-   offers displayed in the legend now use a branded colour, instead of being hardcoded to red

## [0.0.7](https://storage.googleapis.com/ticketswitch/feather/0.0.7/feather.min.js) - 2017-09-06

### Added

-   added a protection against invalid messages, caused by external libraries/browser extensions

### Fixed

-   fixed some IE-specific errors

## [0.0.6](https://storage.googleapis.com/ticketswitch/feather/0.0.6/feather.min.js) - 2017-09-06

### Fixed

-   Fixed a syntax error


main.js
var colors = {};
var domain = getQueryStringParam(window.location.href, 'domain');
var username = getQueryStringParam(window.location.href, 'username');
var password = getQueryStringParam(window.location.href, 'password');
var token = getQueryStringParam(window.location.href, 'token');
var eventID = getQueryStringParam(window.location.href, 'event');
var chart;
var availability;
var basket;

function authAndRedirectDemo() {
  var url = "https://b2b.ingresso.co.uk/api/b2b/";
  var user = "demo";
  var password = "demopass";
  $.ajax
  ({
    type: "GET",
    url: url,
    dataType: 'json',
    headers: {
      "Authorization": "Basic " + btoa(user + ":" + password),
      "Content-Type": "application/json"
    },
    success: function (response, status, info) {
      var token = info.getResponseHeader('X-B2B-Token');
      $("#auth-button").hide();
      createChart("7AB", token, "b2b.ingresso.co.uk", "7AB-4");
    }
  });
};

$('button#select-perf').click(selectNewPerf);
function selectNewPerf(event) {
  var perfID = $('input#perf-id').val();
  chart.selectPerformance(perfID);
}

$('button#add-seat').click(selectSeats);
function selectSeats(selectSeats) {
  var seatID = $('input#seat-id').val();
  chart.selectSeats(seatID);
}

function createChart(eventID, token, domain, performance) {
  if (!token) { token = getQueryStringParam(window.location.href, 'token'); }
  if (!eventID) { eventID = getQueryStringParam(window.location.href, 'event'); }
  if (!domain) { domain = getQueryStringParam(window.location.href, 'domain'); }
  if (!performance) { performance = getQueryStringParam(window.location.href, 'perf'); }

  // Create the chart widget
  // widget configuration
  var chartConfig = {
    eventID: eventID,
    token: token,
    selector: '#chart-container',
    silenceWarnings: false,
    preloaderColor: '#EC008C',
    domain: domain,
    useHTTPS: true,
    hasCustomLegend: true,
  }

  chartConfig.perfID = performance;

  // initialising the widget
  chart = new IngressoSeatingPlan();

  // subscribing to events
  chart.onAddSeat = addSeat;
  chart.onRemoveSeat = removeSeat;
  chart.onEmptyBasket = emptyBasket;
  chart.onSeatsReserved = seatsReserved;
  chart.onGoToCheckout = goToCheckout;
  chart.onNewAvailabilityData = onNewAvailabilityData;
  chart.onNewLegendColors = onNewLegendColors;
  chart.onReserveStopped = onReserveStopped;

  // setting a custom color scheme
  chart.setPreloader({
    html: '<div class="seating-plan-widget-preloader"><span class="rect rect1 bg-color-1"></span><span class="rect rect2 bg-color-1"></span><span class="rect rect3 bg-color-1"></span><span class="rect rect4 bg-color-1"></span><span class="rect rect5 bg-color-1"></span></div>',
    css: '.seating-plan-widget-preloader-container{display: none; position:absolute;width:100%;height:100%;background-color:rgba(255,0,0,.5);user-select:none;top:0;left:0}.seating-plan-widget-preloader{width:300px;height:40px;text-align:center;font-size:10px;position:absolute;left:calc(50% - 150px);top:calc(50% - 30px)}.seating-plan-widget-preloader .rect{background-color:#333;height:100%;width:6px;display:inline-block;margin-right:3px;animation:ing-stretchdelay 1.2s infinite ease-in-out}.seating-plan-widget-preloader .rect2{animation-delay:-1.1s}.seating-plan-widget-preloader .rect3{animation-delay:-1s}.seating-plan-widget-preloader .rect4{animation-delay:-.9s}.seating-plan-widget-preloader .rect5{animation-delay:-.8s}@keyframes ing-stretchdelay{0%,100%,40%{transform:scaleY(.4)}20%{transform:scaleY(1)}}',
  });
  chart.changeColorScheme(['#462446', '#B05F6D', '#EB6B56', '#FFC153', '#47B39D', '#CDCBA6', '#008891', '#00587A', '#ff0000', '#0F3057']);

  chart.showControls();
  chart.disableScrollToZoom();
  $('#checkbox-controls').attr('checked', 'true');

  chart.showLegend();
  $('#checkbox-legend').attr('checked', 'true');

  chart.init(chartConfig);

  // using custom UI elements to control the widget
  $('#zoom-controls #zoom-in').click(chart.zoomIn);
  $('#zoom-controls #zoom-out').click(chart.zoomOut);
  $('#zoom-controls #reset-chart').click(chart.resetChart);
  $('#zoom-controls #show-chart').click(chart.show);
  $('#zoom-controls #hide-chart').click(chart.hide);
  $('#zoom-controls #change-legend').click(changeLegend);
  $("#basket>button#checkout").click(onCheckoutButton);
  $("#basket .empty-basket").click(onEmptyBasketButton);
  $("#checkout-modal #release").click(releaseReservation);
};

function changeLegend() {
  var newLegend = JSON.parse(JSON.stringify(availability.legend));
  newLegend = newLegend.map(function(legendItem, index) {
    legendItem.original_seatprice /= 2;
    legendItem.original_surcharge /= 2;
    legendItem.seatprice /= 2;
    legendItem.surcharge /= 2;
    legendItem.price /= 2;
    legendItem.savingsMessage = 'legend ' + index;
    return legendItem;
  });

  availability.legend = newLegend;
  chart.setLegend(newLegend);
  updateBasket();
}

if (!domain) {
  var button = $('<button/>', {
    text: "Use Ingresso B2B Example",
    id: "auth-button",
    click: authAndRedirectDemo,
  });
  $("#chart-container").append(button);
}
else {
  createChart();
}

function addSeat(data) {
  $('#response-json #content').html(syntaxHighlight(data));
  $('#response-json #event-name').text("addSeat");
  updateBasket(data.basket);
}


function updateBasket(newBasket) {
  if(newBasket) basket = newBasket;
  if(!basket) return;

  $('#basket .seats').empty();
  var total = 0;
  basket.seats.forEach(function(seat) {
    var seatPrice = availability.legend[seat.legend].seatprice + availability.legend[seat.legend].surcharge;
    total += seatPrice;
  });

  basket.seats.forEach(function(seat) {
    var seatElement = $("<p class='item' id='" + seat.uuid + "'> </p>");
    seatElement.append("<span class='color-code' style='border-color: " + colors[seat.legend].selected + "; background-color:" + colors[seat.legend].normal + "'></span>")
    seatElement.append("<span class='description'>" + availability.seat_blocks[seat.seat_block].desc + "</span>")
    seatElement.append("<span class='price'>" + getFormattedPrice(availability.legend[seat.legend].seatprice + availability.legend[seat.legend].surcharge, availability.currency) + "</span>")
    seatElement.append("<span class='remove-seat' data-seat-id='"+seat.uuid+"'><i class='fa fa-trash'></i></span>")
    seatElement.append("<br>");
    seatElement.append("<span style='margin-left: 18px;' class='id'> Seat: " + seat.seat_id + "</span>")
    $('#basket .seats').append(seatElement);
  });
  if(basket.seats.length == 0) {
    $("#basket .total #label").text("Your basket is currently empty");
    $('#basket .total').removeClass("with-seats");
    $("#basket>button#checkout").css('display','none');
    $("#basket .empty-basket").css('display','none');
    $("#basket .total #value").css('opacity', '0');
  } else {
    $("#basket .total #value").css('opacity', '1');
    $("#basket .total #value").text(getFormattedPrice(total, availability.currency));
    $("#basket .total #label").text('Total:');
    $('#basket .total').removeClass("with-seats");
    $('#basket .total').addClass("with-seats");
    $("#basket>button#checkout").css('display','block');
    $("#basket .empty-basket").css('display','block');
  }
  $('.item .remove-seat').unbind('click').bind('click', onRemoveSeatClick);
}

function onRemoveSeatClick(event) {
  var seatUUID = $(event.target).parent().attr('data-seat-id');
  chart.removeSeat(seatUUID);
}

function removeSeat(data) {
  $('#response-json #event-name').text("removeSeat");
  $('#response-json #content').html(syntaxHighlight(data));
  updateBasket(data.basket);
}
function emptyBasket(data) {
  $('#response-json #event-name').text("emptyBasket");
  $('#response-json #content').html(syntaxHighlight(data));
  updateBasket(data.basket);
}


function onNewAvailabilityData(data) {
  availability = data.availability;
}

function onNewLegendColors(data) {
  colors = data.colors;
}

function onReserveStopped() {
  $('button#checkout .preloader').hide();
  $('button#checkout .text').show();
  $('button#checkout').removeClass('disabled');
}

function goToCheckout(data) {
  $('#response-json #content').html(syntaxHighlight(data));
  $('#response-json #event-name').text("goToCheckout");
  $('button#checkout .preloader').hide();
  $('button#checkout .text').show();
  $('button#checkout').removeClass('disabled');
  $("#checkout-modal").modal('show');
  $('#checkout-modal .total').html($('#basket .total').html());
  $('#checkout-modal .transaction-uuid').html('Transaction uuid: </br>' + data.transaction_uuid);
  $('#checkout-modal .seats').html($('#basket .seats').html());
}
function seatsReserved(data) {
  $('#response-json #content').html(syntaxHighlight(data.data));
  $('#response-json #event-name').text("reserve");
  goToCheckout(data);
}

$("#basket .total #value").css('opacity', '0');
$("#basket .total #label").text("Your basket is currently empty");

$("#checkbox-legend").change(onCheckboxLegendChanged);
function onCheckboxLegendChanged(e) {
  if(e.target.checked) {
    chart.showLegend();
  } else {
    chart.hideLegend();
  }
}

$("#checkbox-controls").change(onCheckboxControlsChanged);
function onCheckboxControlsChanged(e) {
  if(e.target.checked) {
    chart.showControls();
  } else {
    chart.hideControls();
  }
}

$("#select-color-scheme").change(onSelectColorSchemeChanged);
function onSelectColorSchemeChanged(e) {
  var option = e.target.selectedIndex;

  if(option == 0) { // default
    chart.changeColorScheme(['#f95e5e', '#FFA726', '#42A5F5', '#fcabc7', '#8febf7', '#55cc5b', '#ae87f2', '#efdc32', '#018930','#a37d70', '#455A64', '#7B1FA2', '#fc9292','#aaaaaa','#fcc18a', '#66FF99']);
  } if(option == 1) { // pastel
    chart.changeColorScheme(['#462446', '#B05F6D', '#EB6B56', '#FFC153', '#47B39D', '#CDCBA6', '#008891', '#00587A', '#ff0000', '#0F3057']);
  } else if (option == 2){ // bold
    chart.changeColorScheme(['#8ea1b3', '#101415', '#ff2600', '#ed5165', '#839496', '#6ec14c', '#d7a302', '#0084cc', '#b433e1']);
  } else if (option == 3){ // grayscale
    chart.changeColorScheme(['#444', '#555', '#666', '#777', '#888', '#999', '#aaa', '#bbb', '#ccc', '#aaa', '#444', '#aaa', '#444', '#aaa', '#444', '#aaa']);
  } else if(option == 4) { // sea shells
    chart.changeColorScheme(['#f95e5e', '#071a24', '#265b76', '#60b2c2', '#dd9a8a', '#005cbb', '#c7e8f2', '#e5c7f2', '#f95e5e', '#071a24', '#265b76', '#60b2c2', '#dd9a8a', '#005cbb', '#c7e8f2', '#e5c7f2']);
  }
}


function releaseReservation() {
  chart.release();
  $("#checkout-modal").modal('hide');
}

function onCheckoutButton() {
  chart.reserve();
  $('button#checkout').addClass('disabled');
  $('button#checkout .preloader').show();
  $('button#checkout .text').hide();
}

function onEmptyBasketButton() {
  chart.emptyBasket();
}

/**
 These utility functions are not necessary in order to make full use
 of the widget, but proved useful for the demo and are production-ready
**/

function getPreSymbol(currencyObj) {
  if(!currencyObj || !currencyObj[Object.keys(currencyObj)[0]]) return "";
  var currency = currencyObj[Object.keys(currencyObj)[0]];
  return currency.pre_symbol;
}

function getPostSymbol(currencyObj) {
  if(!currencyObj || !currencyObj[Object.keys(currencyObj)[0]]) return "";
  var currency = currencyObj[Object.keys(currencyObj)[0]];
  return currency.post_symbol;
}

function getFormattedPrice(price, currencyObj, roundIfOverHundred) {
  if(!currencyObj) return;
  var preSymbol = currencyObj.pre_symbol || "";
  var postSymbol = currencyObj.post_symbol || "";
  if(roundIfOverHundred && parseFloat(price) > 100) price = parseInt(price);
  return preSymbol + parseFloat(price).toFixed(2) + postSymbol;
}

function getQueryStringParam(url, paramName) {
	var valToReturn = null;
	try {
		valToReturn = url.split(paramName+'=')[1].split('&')[0];
		if(valToReturn.indexOf('#') != -1) {
			valToReturn = valToReturn.split('#')[0];
		}
	}
	catch(err) {
		// it will fail when the parameter doesn't exist; this is normal
	}
	return valToReturn;
}

function syntaxHighlight(json) {
    if (typeof json != 'string') {
         json = JSON.stringify(json, undefined, 2);
    }
    json = json.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
    return json.replace(/("(\\u[a-zA-Z0-9]{4}|\\[^u]|[^\\"])*"(\s*:)?|\b(true|false|null)\b|-?\d+(?:\.\d*)?(?:[eE][+\-]?\d+)?)/g, function (match) {
        var cls = 'number';
        if (/^"/.test(match)) {
            if (/:$/.test(match)) {
                cls = 'key';
            } else {
                cls = 'string';
            }
        } else if (/true|false/.test(match)) {
            cls = 'boolean';
        } else if (/null/.test(match)) {
            cls = 'null';
        }
        return '<span class="' + cls + '">' + match + '</span>';
    });
}
